src/Workspaces/Core/Portable/FindSymbols/FindReferences/DependentTypeFinder.cs

```diff
@@ -656,58 +654,131 @@ internal static class DependentTypeFinder
+                        sourceTypeImmediatelyMatches, cancellationToken).ConfigureAwait(false);
+                }
 
                 // Clear out the information about the types we're looking for.  We'll
                 // fill these in if we discover any more types that we need to keep searching
                 // for.
                 typesToSearchFor.Clear();
-                inheritanceQuery.TypeNames.Clear();
 
-                foreach (var task in tasks)
+                foreach (var derivedType in result)
                 {
-                    var result = await task.ConfigureAwait(false);
-                    foreach (var derivedType in result)
+                    if (finalResult.Add(derivedType))
                     {
-                        if (finalResult.Add(derivedType))
+                        if (transitive && shouldContinueSearching(derivedType.Symbol))
                         {
-                            if (transitive && shouldContinueSearching(derivedType.Symbol))
-                            {
-                                typesToSearchFor.Add(derivedType);
-                            }
+                            typesToSearchFor.Add(derivedType);
                         }
                     }
+                }
 
-                    s_setPool.ClearAndFree(result);
+                s_setPool.ClearAndFree(result);
+            }
+        }
+
+        private static readonly ConditionalWeakTable<Project, AsyncLazy<ProjectIndex>> s_projectToIndex = 
+            new ConditionalWeakTable<Project, AsyncLazy<ProjectIndex>>();
+
+        private static Task<ProjectIndex> GetProjectIndexAsync(
+            Project project, CancellationToken cancellationToken)
+        {
+            if (!s_projectToIndex.TryGetValue(project, out var lazyIndex))
+            {
+                lazyIndex = s_projectToIndex.GetValue(
+                    project, p => new AsyncLazy<ProjectIndex>(c => CreateIndex(p, c), cacheResult: true));
+            }
+
+            return lazyIndex.GetValueAsync(cancellationToken);
+        }
+
+        private static async Task<ProjectIndex> CreateIndex(Project project, CancellationToken cancellationToken)
+        {
+            var classesThatMayDeriveFromSystemObject = new MultiDictionary<Document, DeclaredSymbolInfo>();
+            var valueTypes = new MultiDictionary<Document, DeclaredSymbolInfo>();
+            var enums = new MultiDictionary<Document, DeclaredSymbolInfo>();
+            var delegates = new MultiDictionary<Document, DeclaredSymbolInfo>();
+            var namedTypes = new MultiDictionary<string, (Document, DeclaredSymbolInfo)>(
+                project.LanguageServices.GetService<ISyntaxFactsService>().StringComparer);
+
+            foreach (var document in project.Documents)
+            {
+                var syntaxTreeIndex = await document.GetSyntaxTreeIndexAsync(cancellationToken).ConfigureAwait(false);
```

Comment:
So this would be a little difficult to parallelize.  Because this is the code generating the 'merged' index.  We're basically doing nothing but writes to a shared datastruture, and that will likely be quite contorted/difficult to do in a parallel manner .